HW2/P4/driver.py

```diff
@@ -27,6 +28,77 @@ def py_median_3x3(image, iterations=10, num_threads=1):
 
     return tmpA
 
+
+def py_median_3x3_sync(image, iterations=10, num_threads=1):
+    ''' repeatedly filter with a 3x3 median '''
+    tmpA = image.copy()
+    tmpB = np.empty_like(tmpA)
+
+    for i in range(iterations):
+        threads = []
+        for j in xrange(num_threads):
+            t = threading.Thread(target=filtering.median_3x3, args=(tmpA, tmpB,
+                                 j, num_threads))
+            t.start()
+            threads.append(t)
+        for t in threads:
+            t.join()
+        # swap direction of filtering
+        tmpA, tmpB = tmpB, tmpA
+
+    return tmpA
+
+
+def py_median_3x3_async(image, iterations=10, num_threads=1):
+    ''' repeatedly filter with a 3x3 median '''
+    tmpA = image.copy()
+    tmpB = np.empty_like(tmpA)
+
+    # List of event for each iteration & thread
+    events = []
+    for i in xrange(iterations):
+        events_list = [threading.Event()]
```

Comment:
I used a very similar implementation with Event() objects.  Glad to see I wasn't alone!